### PR TITLE
[LC-511] verify next_leader field in v0.3 block with each node's counter

### PR DESCRIPTION
**Pull request by windies21**

### CODE REVIEW

File: loopchain/blockchain/blockchain.py [hunk 1]
```diff
@@ -98,6 +101,35 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
         for tx_version, tx_hash_version in channel_option.get("hash_versions", {}).items():
             self.__tx_versioner.hash_generator_versions[tx_version] = tx_hash_version
 
+    @property
+    def leader_made_block_count(self):
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py]:
```suggestion
    def leader_made_block_count(self) -> int:
```

File: loopchain/blockchain/blockchain.py [hunk 2]
```diff
@@ -98,6 +101,35 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
...
+        if self.__last_block:
+            return self.__last_leader_made_block_count[self.__last_block.header.peer_id]
+        return -1
+
+    @property
+    def my_made_block_count(self):
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py]:
```suggestion
    def my_made_block_count(self) -> int:
```

File: loopchain/blockchain/blockchain.py [hunk 3]
```diff
@@ -98,6 +101,35 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
...
+        return self.__last_leader_made_block_count[ChannelProperty().peer_address]
+
+    def _up_leader_made_block_count(self):
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py]:
```suggestion
    def _up_leader_made_block_count(self) -> int:
```

File: loopchain/blockchain/blockchain.py [hunk 4]
```diff
@@ -98,6 +101,35 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
...
+        self.__last_leader_made_block_count[self.__last_block.header.peer_id] += 1
+
+    def reset_leader_made_block_count(self):
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py]:
```suggestion
    def reset_leader_made_block_count(self) -> int:
```

File: loopchain/blockchain/blockchain.py [hunk 5]
```diff
@@ -98,6 +101,35 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
...
+        self.__last_leader_made_block_count.clear()
+
+    def get_next_leader(self):
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py]: I think the return value is ExternalAddress (I'm not sure), but PeerManager.peer_id set its value as str.

I suggest below. but setting return type as ExternalAddress could breach style convention on IDE.
```suggestion
    def get_next_leader(self) -> ExternalAddress:
```

File: loopchain/channel/channel_service.py [hunk 6]
```diff
@@ -196,6 +196,7 @@ def cleanup(self):
         ChannelProperty().rest_target = kwargs.get('rest_target')
         ChannelProperty().radio_station_target = kwargs.get('rs_target')
         ChannelProperty().peer_id = kwargs.get('peer_id')
+        ChannelProperty().peer_address = ExternalAddress.fromhex_address(ChannelProperty().peer_id)
```
Comment (laurenceyoon) [inline loopchain/channel/channel_service.py]: How about integrating `peer_id` to `peer_address` and change usages of `ChannelProperty().peer_id` to `ChannelProperty().peer_address.hex_hx()`?

File: loopchain/baseservice/node_subscriber.py [hunk 7]
```diff
@@ -125,7 +125,7 @@ def __del__(self):
             vote = BlockVotes.deserialize_votes(votes_serialized)
         else:
             vote = None
-        blockchain = ObjectManager().channel_service.block_manager.get_blockchain()
+        blockchain = ObjectManager().channel_service.block_manager.blockchain
```
Comment (metalg0su) [inline loopchain/baseservice/node_subscriber.py]: How about accessing `BlockManager.blockchain' variable as `@property?`, not directly accessing it?

File: loopchain/peer/block_manager.py [hunk 8]
```diff
@@ -61,7 +61,7 @@ def __init__(self, name: str, channel_service, peer_id, channel_name, store_iden
 
         self.__txQueue = AgingCache(max_age_seconds=conf.MAX_TX_QUEUE_AGING_SECONDS,
                                     default_item_status=TransactionStatusInQueue.normal)
-        self.__blockchain = BlockChain(channel_name, store_identity, self)
+        self.blockchain = BlockChain(channel_name, store_identity, self)
```
Comment (metalg0su) [inline loopchain/peer/block_manager.py]: https://github.com/icon-project/loopchain/pull/274/files#r315968699

File: loopchain/blockchain/blockchain.py [hunk 9]
```diff
@@ -98,6 +111,60 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
...
+    def leader_made_block_count(self) -> int:
+        if self.__last_block:
+            return self.__made_block_count[self.__last_block.header.peer_id]
+        return -1
+
+    @property
+    def my_made_block_count(self) -> int:
+        return self.__made_block_count[ChannelProperty().peer_address]
+
+    def _up_made_block_count(self, block: Block) -> None:
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py]: just suggestion, if you intended increasing the value of `__made_block_count`.
it is also go along with `reset_leader_made_block_count` and other methods, which are consisted of "verb" and its "object".

```suggestion
    def _increase_made_block_count(self, block: Block) -> None:
```

File: loopchain/blockchain/blockchain.py [hunk 10]
```diff
@@ -98,6 +111,60 @@ def __init__(self, channel_name=None, store_id=None, block_manager=None):
...
+    def get_expected_generator(self, unconfirmed_block: Block) -> str:
+        """get expected generator to vote unconfirmed block
+
+        :return: expected generator's id by made block count.
+        """
+
+        peer_manager = ObjectManager().channel_service.peer_manager
+        if self.__made_block_count[unconfirmed_block.header.peer_id] > conf.MAX_MADE_BLOCK_COUNT:
+            return peer_manager.get_next_leader_peer(unconfirmed_block.header.peer_id).peer_id
+        return unconfirmed_block.header.peer_id.hex_hx()
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py]: How about pull of `unconfirmed_block.header.peer_id` as variable?
It is repeated 3 times!

File: loopchain/peermanager/peer_loader.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            logging.debug(f"reps by c-rep root hash: {reps}")
+            for order, rep_info in enumerate(reps, 1):
+                peer = Peer(rep_info["address"], rep_info["p2pEndpoint"], order=order)
+                peer_manager.add_peer(peer)
+            return
+
+        response = rest_stub.call("GetChannelInfos")
+        reps: list = response['channel_infos'][ChannelProperty().name].get('peers')
+        for peer_info in reps:
+            peer_manager.add_peer(peer_info)
```
Comment (metalg0su) [inline loopchain/peermanager/peer_loader.py]: all of methos in `PeerLoader` is staticmethod, and it has file name as `peer_loader.py`.
I don't know what is better, but I carefully suggest that all this helper methods do not need its class `PeerLoader`, module `peer_loader.py` is enough.

File: loopchain/blockchain/blockchain.py [hunk 12]
```diff
@@ -49,6 +50,16 @@ class NID(Enum):
     unknown = "0x3"
 
 
+class MadeBlockCounter(Counter):
+    def __str__(self):
```
[the remark this answers is not in the document]
Comment (windies21) [reply loopchain/blockchain/blockchain.py]: This class will be removed. This class is for debugging purposes only.

File: loopchain/channel/channel_service.py (hunk 6 above)
Comment (windies21) [reply loopchain/channel/channel_service.py]: More discussion is needed. It can be changed in other PR or Not.

**Pull request merged**